### PR TITLE
parser.act: Use a hash table for token interning in the parser.

### PR DESCRIPTION
**Pull request by silentbicycle**
The parser's token interning dominates everything else when profiling
large test data sets, because previously it was walking an entire linked
list of tokens to find matches, which is quadratic on input size.

Switch to an add-only hash table with separate chaining for the tokens,
and grow whenever any of the chains are longer than MAX_CHAIN_LENGTH.

This uses FNV-1a for hashing, which is in the public domain.

I will update the generated parser code in a separate commit.

### CODE REVIEW

File: src/fsm/parser.act [hunk 1]
```diff
@@ -44,23 +44,32 @@
 	typedef char *      string;
 	typedef fsm_state_t state;
 
+	struct act_statelist {
+		char *id;
+		fsm_state_t state;
+		struct act_statelist *next;
+	};
```
Comment (katef) [inline src/fsm/parser.act:51]: I wonder if we shouldn't get rid of the idea of a string `id` for the fsm(1) syntax entirely. But maybe not, writing `1a`, `1b` etc is convenient for some tests, especially when the expected output is combined from two FSMs.

File: src/fsm/parser.act [hunk 2]
```diff
@@ -90,6 +99,58 @@
...
+
+	static void
+	grow_state_hash_table(struct act_stateset *htab)
+	{
+		const size_t ncount = 2 * htab->bucket_count;
+		struct act_statelist **nbuckets = xcalloc(ncount, sizeof(*nbuckets));
+		size_t max_chain_length = 0;
+		size_t old_i;
+		unsigned mask = ncount - 1;
+		if (nbuckets == NULL) {
```
Comment (katef) [inline src/fsm/parser.act:127]: This can never happen now (because xcalloc will exit)

File: src/fsm/parser.act [hunk 3]
```diff
@@ -336,7 +413,10 @@
 		lx->clear      = lx_dynclear;
 		lx->free       = lx_dynfree;
 
-		act_state_s.sl = NULL;
+		act_state_s.states.buckets = calloc(DEF_BUCKETS, sizeof(act_state_s.states.buckets));
```
Comment (katef) [inline src/fsm/parser.act:416]: `xcalloc` here too :)

File: src/fsm/parser.act (hunk 1 above)
Comment (silentbicycle) [reply src/fsm/parser.act:51]: I'd prefer to keep that flexible, it's been helpful when generating test data too.

File: src/fsm/parser.act (hunk 2 above)
Comment (silentbicycle) [reply src/fsm/parser.act]: Fixed, amended.

File: src/fsm/parser.act (hunk 3 above)
Comment (silentbicycle) [reply src/fsm/parser.act]: Fixed, amended.

**Pull request merged**